keras/applications/nasnet.py

```diff
@@ -0,0 +1,692 @@
+"""NASNet-A models for Keras
+
+NASNet refers to Neural Architecture Search Network, a family of models
+that were designed automatically by learning the model architectures
+directly on the dataset of interest.
+
+Here we consider NASNet-A, the highest performance model that was found
+for the CIFAR-10 dataset, and then extended to ImageNet 2012 dataset,
+obtaining state of the art performance on CIFAR-10 and ImageNet 2012.
+Only the NASNet-A models, and their respective weights, which are suited
+for ImageNet 2012 are provided.
+
+The below table describes the performance on ImageNet 2012:
+------------------------------------------------------------------------------------
+      Architecture       | Top-1 Acc | Top-5 Acc |  Multiply-Adds |  Params (M)
+------------------------------------------------------------------------------------
+|   NASNet-A (4 @ 1056)  |   74.0 %  |   91.6 %  |       564 M    |     5.3        |
+|   NASNet-A (6 @ 4032)  |   82.7 %  |   96.2 %  |      23.8 B    |    88.9        |
+------------------------------------------------------------------------------------
+
+Weights obtained from the official Tensorflow repository found at
+https://github.com/tensorflow/models/tree/master/research/slim/nets/nasnet
+
+# References:
+ - [Learning Transferable Architectures for Scalable Image Recognition]
+    (https://arxiv.org/abs/1707.07012)
+
+Based on the following implementations:
+ - [TF Slim Implementation]
+   (https://github.com/tensorflow/models/blob/master/research/slim/nets/nasnet/nasnet.py)
+ - [TensorNets implementation]
+   (https://github.com/taehoonlee/tensornets/blob/master/tensornets/nasnets.py)
+"""
+from __future__ import print_function
+from __future__ import absolute_import
+from __future__ import division
+
+import warnings
+
+from ..models import Model
+from ..layers import Input
+from ..layers import Activation
+from ..layers import Dense
+from ..layers import Dropout
+from ..layers import BatchNormalization
+from ..layers import MaxPooling2D
+from ..layers import AveragePooling2D
+from ..layers import GlobalAveragePooling2D
+from ..layers import GlobalMaxPooling2D
+from ..layers import Conv2D
+from ..layers import SeparableConv2D
+from ..layers import ZeroPadding2D
+from ..layers import Cropping2D
+from ..layers import concatenate
+from ..layers import add
+from ..regularizers import l2
+from ..utils.data_utils import get_file
+from ..engine.topology import get_source_inputs
+from ..applications.imagenet_utils import _obtain_input_shape
+from ..applications.inception_v3 import preprocess_input
+from ..applications.imagenet_utils import decode_predictions
+from .. import backend as K
+
+_BN_DECAY = 0.9997
+_BN_EPSILON = 1e-3
+
+NASNET_MOBILE_WEIGHT_PATH = "https://github.com/titu1994/Keras-NASNet/releases/download/v1.0/NASNet-mobile.h5"
```

Comment:
Use `'` as string delimiter, for consistency. This applies everywhere in the file.